graphbolt/src/fused_csc_sampling_graph.cc

```diff
@@ -607,27 +607,37 @@ FusedCSCSamplingGraph::SampleNeighborsImpl(
 }
 
 c10::intrusive_ptr<FusedSampledSubgraph> FusedCSCSamplingGraph::SampleNeighbors(
-    const torch::Tensor& nodes, const std::vector<int64_t>& fanouts,
+    torch::optional<torch::Tensor> nodes, const std::vector<int64_t>& fanouts,
     bool replace, bool layer, bool return_eids,
     torch::optional<std::string> probs_name) const {
-  torch::optional<torch::Tensor> probs_or_mask = torch::nullopt;
-  if (probs_name.has_value() && !probs_name.value().empty()) {
-    probs_or_mask = this->EdgeAttribute(probs_name);
-  }
+  auto probs_or_mask = this->EdgeAttribute(probs_name);
 
-  if (!replace && utils::is_on_gpu(nodes) &&
-      utils::is_accessible_from_gpu(indptr_) &&
-      utils::is_accessible_from_gpu(indices_) &&
-      (!probs_or_mask.has_value() ||
-       utils::is_accessible_from_gpu(probs_or_mask.value())) &&
-      (!type_per_edge_.has_value() ||
-       utils::is_accessible_from_gpu(type_per_edge_.value()))) {
+  // If nodes does not have a value, then we expect all arguments to be resident
+  // on the GPU. If nodes has a value, then we expect them to be accessible from
+  // GPU. This is required for the dispatch to work when CUDA is not available.
+  if (((!nodes.has_value() && utils::is_on_gpu(indptr_) &&
+        utils::is_on_gpu(indices_) &&
+        (!probs_or_mask.has_value() ||
+         utils::is_on_gpu(probs_or_mask.value())) &&
+        (!type_per_edge_.has_value() ||
+         utils::is_on_gpu(type_per_edge_.value()))) ||
+       (nodes.has_value() && utils::is_on_gpu(nodes.value()) &&
+        utils::is_accessible_from_gpu(indptr_) &&
+        utils::is_accessible_from_gpu(indices_) &&
+        (!probs_or_mask.has_value() ||
+         utils::is_accessible_from_gpu(probs_or_mask.value())) &&
+        (!type_per_edge_.has_value() ||
+         utils::is_accessible_from_gpu(type_per_edge_.value())))) &&
+      !replace) {
     GRAPHBOLT_DISPATCH_CUDA_ONLY_DEVICE(
         c10::DeviceType::CUDA, "SampleNeighbors", {
           return ops::SampleNeighbors(
               indptr_, indices_, nodes, fanouts, replace, layer, return_eids,
               type_per_edge_, probs_or_mask);
         });
+  } else if (!nodes.has_value()) {
+    // Initialize the nodes tensor for the CPU execution.
+    nodes = torch::arange(NumNodes(), indices_.options());
```

Comment:
Shall we just throw error for cpu cases?

Reply:
That will complicate the tests a little. Would need to make a special case for the CPU.

Reply:
I prefer keeping it this way but if you think it is better to throw an error for the CPU, I can do that before merging.

Reply:
The disk case is not the same as the GPU. Let me throw an error.

Reply:
Added a check for the CPU. CPU requires nodes to be passed. We can change it later if needed.